### DISCUSSION TITLE
Correct interface with LPSOLVE?

### DISCUSSION THREAD

**Discussion by mfry-ll** [yalmip/YALMIP#995] [Possible bug or missing functionality] [answered]
I'm trying to use [LPSOLVE](https://sourceforge.net/projects/lpsolve/files/lpsolve/5.5.2.9/lp_solve_5.5.2.9_MATLAB_exe_ux64.tar.gz/download), but it appears that YALMIP is calling the `lp_solve.m` incorrectly (or their API has just changed in their more recent versions):
```
Aeq = eye(3);
beq = rand(3, 1);
lb = zeros(3, 1);
x = sdpvar(3, 1);
ct = [Aeq * x == beq, lb <= x];
options = sdpsettings('solver', 'lpsolve', 'verbose', false);
optimize(ct, [], options)                                    % Throws an error
options.solver = 'mosek';
optimize(ct, [], options)                                    % No issues
```
I want to avoid relying on `linprog` or `mosek` because of licensing.

Also, even though this is failing, `yalmiptest('lpsolve')` reports that it is solving the linear and MILP problems (but really, they're hitting the same error as above).

**Comment by johanlofberg** [comment]
I doubt there has been any changes as that project has been pretty dead for some 15 years.

You will have to turn on debug so you can report what actually happens.

**Comment by PiotrBalik** [comment, marked as answer]
Assuming you have working installation.
You have to replace `lp_solve` with `mxlpsolve` (simply ctrl+h, whole word) in two files: 
`solvers/calllpsolve.m`
`solvers/create_lp_solve_model.m`

Working on windows, you've attached linux link, but let's hope yalmip calls are the same

**Reply by johanlofberg** [reply]
Note that YALMIPs support for lpsolve is intended to be the version shipped in OPTI Toolbox https://yalmip.github.io/solver/lpsolve/, which works as expected. 

lpsolve support will most likely be removed in a not too distant future. The current develop branch has already pruned a large amount of effectively dead solvers.

**Reply by PiotrBalik** [reply]
Taking OPTI into account I think there's even better and simpler solution:
@mfry-ll rename `mxlpsolve.mexa64` to `lp_solve.mexa64`, replace it in OPTI installation path and it will work without any changes to yalmip structure

Regarding lpsolve being dead, it has been updated in unpredictable way
![obraz](https://user-images.githubusercontent.com/59842832/133613151-12adc5ec-93b8-42c0-bd35-12568e58e7fa.png)

**Reply by johanlofberg** [reply]
Ah. But then dead is also takes usage among YALMIP users into account. I know there is 1 user (soon) but beyond that I do not know...However, if the interface works I don't have to kill it

**Reply by mfry-ll** [reply]
> @mfry-ll rename `mxlpsolve.mexa64` to `lp_solve.mexa64`, replace it in OPTI installation path and it will work without any changes to yalmip structure

Thanks @PiotrBalik , that works.

@johanlofberg, before trying that, I attempted to use the version shipped with OPTI toolbox, but their documentation on the provided version is contradictory 
![image](https://user-images.githubusercontent.com/89883463/133656218-585af684-2065-4285-90c2-5ad6df1a3637.png)
[source](https://github.com/jonathancurrie/OPTI/blob/master/Solvers/Source/lpsolve/lpsolve.c)
[changelog](https://github.com/jonathancurrie/OPTI/blob/master/Solvers/Documentation%20%2B%20Licenses/lpsolve/changes)
All listing different versions of LPSOLVE.

I want to provide an installation script for LPSOLVE, which, via @PiotrBalik's solution, would involve downloading, unzipping, and renaming a file (not too difficult). However, for dependency consistency, it'd be important to know if @johanlofberg is planning on updating yalmip to interface with the more recent version of LPSOLVE (i.e., replacing calls to `lp_solve()` with calls to `mxlpsolve`), or if you intend to keep yalmip aligned with the interface of the older version.

I'm not sure if planning to "follow whatever OPTI toolbox's version is" would be a good solution, since they don't seem to report a consistent version. I also tried to interface YALMIP with one of the versions claimed by OPTI toolbox (LPSOLVE 5.5.2.0), and I got the same errors. Also, it appears OPTI toolbox is exclusive to Windows, so Unix users won't be able to just install OPTI toolbox and call LPSOVLE with yalmip.

**Reply by mfry-ll** [reply]
@PiotrBalik @johanlofberg renaming `mxlpsolve.mexa64` to `lp_solve.mexa64` creates another problem, because LPSOLVE provides both `mxlpsolve.mexa64` and `lp_solve.m`.  By renaming `mxlpsolve.mexa64` to `lp_solve.mexa64`, any native use of LPSOLVE (without calling from YALMIP, as in LPSOLVE's example scripts) will fail, because those make calls to `lp_solve.m`. 

I think the YALMIP fix is relatively easy (change calls from `lp_solve()` to `mxlpsolve`), if you want to elevate this to an issue, I can make a fix and PR it.

**Reply by PiotrBalik** [reply]
So we're changing either yalmip solver list or try to work around native call.
Wouldn't it be easier to change native interface `lp_solve.m` to something different, which calls `lp_solve.mexa64` anyway?

Looks like there's pr already. . .

**Reply by johanlofberg** [reply]
lp_solve is is not a yalmip file so it cannot be changed, if that is what you meant. But yeah, fix already cooking

**Reply by mfry-ll** [reply]
Thanks for merging!
